File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    private void optimalSol(List<String> list,StringBuilder curr,int open ,int close,int n){
        if(curr.length()==2*n){
            list.add(curr.toString());
            return;
        }

        if(open < n){
            curr.append('(');
            optimalSol(list,curr,open + 1,close,n);
            curr.deleteCharAt(curr.length()-1);
        }
         if(close < open){
            curr.append(')');
            optimalSol(list,curr,open,close +1 ,n);
            curr.deleteCharAt(curr.length()-1);
        }


    }
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        StringBuilder ans = new StringBuilder();

        optimalSol(result,ans,0,0,n);
        return result;
        
    }
}

File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
private:
    bool isValid(const string& str) {
        int balance = 0;
        for (char ch : str) {
            if (ch == '(')
                balance++;
            else
                balance--;
                if (balance < 0) return false;
        }
        return balance == 0;
    }
     void generateAll(string& curr ,int n,vector<string>& result){
        if(curr.length()==2*n){
            if(isValid(curr)){
                result.push_back(curr);
            }
            return;
        }
        curr.push_back('(');
        generateAll(curr,n,result);
        curr.pop_back();

        curr.push_back(')');
        generateAll(curr,n,result);
        curr.pop_back();
     }

public:
    vector<string> generateParenthesis(int n) {
        vector<string>result;
        string ans="";
        generateAll(ans,n,result);
        return result;
    }
};